04 - Functional Programming/04 - Functional Programming.ipynb

```diff
@@ -0,0 +1,3551 @@
+        "Нека усложим нещата - ако се върнем на първоначалната ни дефиниция за `map` (Функция, която прилага друга функция върху колекция), си задаваме въпроса - какво става, ако имаме повече от една колекция, върху която искаме да приложим функция ? Възможно ли е въобще ? \n",
+        "\n",
+        "Отговора е да - `map` приема една или повече колекции. Единствената особенност е, че функцията която подаваме трябва да може да приеме повече от един аргумент. На по-прост език - ако подадем два списъка на `map`, функцията която ще се приложи върху тях трябва да приема два аргумента."
+      ],
+      "metadata": {
+        "id": "VaA7b3_RTzp-"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "first_numbers = [1, 2, 3, 4]\n",
+        "second_numbers = [5, 6, 7, 8]\n",
+        "\n",
+        "def multiply(a, b):\n",
+        "    return a * b\n",
+        "\n",
+        "print(list(map(multiply, first_numbers, second_numbers)))"
+      ],
+      "metadata": {
+        "colab": {
+          "base_uri": "https://localhost:8080/"
+        },
+        "id": "u4peYrvxULBq",
+        "outputId": "51a5a3cb-bf5a-4806-e3a0-1b4461bb8189"
+      },
+      "execution_count": 41,
+      "outputs": [
+        {
+          "output_type": "stream",
+          "name": "stdout",
+          "text": [
+            "[5, 12, 21, 32]\n"
+          ]
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "Важно е да се отбележи, че резултата от `map` операция винаги е **една** поредица от обекти - дори и да приемем 10 колекции, винаги на изхода ще имаме една. \n",
+        "\n",
+        "**Въпрос:** Как може да \"измамим\" системата, и да върнем повече от една колекция ?"
+      ],
+      "metadata": {
+        "id": "mpij1VPbWQyW"
+      }
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "### Отговор"
+      ],
+      "metadata": {
+        "id": "yIVTK-AEYJLZ"
+      }
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "Няма как да върнем повече от една колекция - това което можем да направим, е да върнем колекция, която съдържа наредени n-торки (tuples)."
+      ],
+      "metadata": {
+        "id": "hnXPTxPYXym_"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "first_numbers = [1, 2, 3, 4]\n",
+        "second_numbers = [5, 6, 7, 8]\n",
+        "\n",
+        "def return_as_tuple(a, b):\n",
+        "    return a, b\n",
+        "\n",
+        "print(list(map(return_as_tuple, first_numbers, second_numbers)))"
+      ],
+      "metadata": {
+        "colab": {
+          "base_uri": "https://localhost:8080/"
+        },
+        "id": "-07cxylmWQDR",
+        "outputId": "52848409-0b1d-4e06-e9ef-57f8748d0365"
+      },
+      "execution_count": 42,
+      "outputs": [
+        {
+          "output_type": "stream",
+          "name": "stdout",
+          "text": [
+            "[(1, 5), (2, 6), (3, 7), (4, 8)]\n"
```

Comment:
I would move the introduction of zip here as a quick plug-in.

Reply:
Agree, changed.